dace/codegen/targets/fpga.py

```diff
@@ -676,6 +851,162 @@ def deallocate_array(self, sdfg, dfg, state_id, node, function_stream,
                          callsite_stream):
         pass  # Handled by destructor
 
+    def compute_kernels(self, state: dace.SDFGState, default_kernel: int = 0):
+        """ Annotates  state nodes to include a `_kernel` ID field.
+            This field is applied to all FPGA maps, tasklets, and library nodes
+            that can be executed in parallel in separate kernels.
+
+            :param state: the state to analyze.
+            :param default_kernel: The Kernel ID to start counting from.
+            :return: a tuple containing the number of kernels and the dependencies among them
+        """
+
+        # TODO:
+        #  - control  this by some dace configuration flag?
+
+        concurrent_kernels = 0  # Max number of kernels
+        sdfg = state.parent
+
+        def increment(kernel_id):
+            if concurrent_kernels > 0:
+                return (kernel_id + 1) % concurrent_kernels
+            return kernel_id + 1
+
+        # Dictionary containing dependencies among kernels:
+        # dependencies[K] = [list of kernel IDs from which K depends]
+        dependencies = dict()
+
+        source_nodes = state.source_nodes()
+        max_kernels = default_kernel
+
+        # First step: assign a different Kernel ID
+        # to each source node which is not an AccessNode
+        for i, node in enumerate(source_nodes):
+            if isinstance(node, nodes.AccessNode):
+                continue
+
+            node._kernel = max_kernels
```

Comment:
Nooo... Please use an auxiliary data structure instead of secret fields on the graph, if at all possible. Can't we just keep a mapping in the codegen?